<h1 align=center> Exploratory Data Analysis

### New York City Taxi trip Records

Exploración de los datos obtenidos del NYC (Taxi & Limusine commission) [https://www.nyc.gov/site/tlc/about/tlc-trip-record-data.page] 
sobre recorridos de taxis amarillos, verdes y for-hire en Nueva York año 2024

In [3]:
import pandas as pd
import numpy as np

### 1. Archivo yellow_tripdata_2024_05.parquet

In [19]:
df1 = pd.read_parquet(r'C:\Users\migue\OneDrive\Escritorio\Datasets Taxi\yellow_tripdata_2024-05.parquet', engine='pyarrow')
df1

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2024-05-01 00:59:15   2024-05-01 01:23:50              1.0   
1               2  2024-04-30 23:58:26   2024-05-01 00:29:42              1.0   
2               2  2024-05-01 00:57:17   2024-05-01 01:14:15              1.0   
3               2  2024-05-01 00:24:47   2024-05-01 00:48:51              1.0   
4               2  2024-05-01 00:11:20   2024-05-01 00:52:10              1.0   
...           ...                  ...                   ...              ...   
3723828         2  2024-05-31 23:45:00   2024-06-01 00:09:00              NaN   
3723829         2  2024-05-31 23:10:52   2024-05-31 23:38:05              NaN   
3723830         2  2024-05-31 23:23:57   2024-05-31 23:38:35              NaN   
3723831         2  2024-05-31 23:44:12   2024-05-31 23:47:44              NaN   
3723832         2  2024-05-31 23:11:10   2024-05-31 23:38:27              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 6.10         1.0                  N           138   
1                11.23         1.0                  N           138   
2                 9.02         1.0                  N           138   
3                 6.53         1.0                  N            87   
4                14.38         1.0                  N           161   
...                ...         ...                ...           ...   
3723828           4.07         NaN               None            88   
3723829           4.81         NaN               None           231   
3723830           2.03         NaN               None           142   
3723831           0.94         NaN               None           141   
3723832           9.25         NaN               None           232   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 145             1        28.20   7.75      0.5        5.00   
1                 249             1        46.40   6.00      0.5        8.72   
2                 170             1        35.90   6.00      0.5       10.57   
3                 133             1        30.30   1.00      0.5        7.06   
4                 165             1        61.80   1.00      0.5        0.00   
...               ...           ...          ...    ...      ...         ...   
3723828           186             0        29.93   0.00      0.5        0.00   
3723829           143             0        30.71   0.00      0.5        0.00   
3723830           140             0        17.15   0.00      0.5        4.23   
3723831           263             0        -1.00   0.00      0.5        0.00   
3723832            42             0        33.65   0.00      0.5        0.00   

         tolls_amount  improvement_surcharge  total_amount  \
0                0.00                    1.0         42.45   
1                0.00                    1.0         66.87   
2                6.94                    1.0         65.16   
3                0.00                    1.0         42.36   
4                0.00                    1.0         66.80   
...               ...                    ...           ...   
3723828          0.00                    1.0         33.93   
3723829          0.00                    1.0         34.71   
3723830          0.00                    1.0         25.38   
3723831          0.00                    1.0          3.00   
3723832          0.00                    1.0         37.65   

         congestion_surcharge  Airport_fee  
0                         0.0         1.75  
1                         2.5         1.75  
2                         2.5         1.75  
3                         2.5         0.00  
4                         2.5         0.00  
...                       ...          ...  
3723828                   NaN          NaN  
3723829                   NaN          NaN  
3723830                   NaN          NaN  
3723831                   N

Observamos 3.7 millones de viajes registrados, con información en 19 columnas que abarca desde tiempo y localización de comienzo y fin del viaje, como distancia, tarifas y cobros extras

In [20]:
df1.isna().sum()

VendorID                      0
tpep_pickup_datetime          0
tpep_dropoff_datetime         0
passenger_count          404666
trip_distance                 0
RatecodeID               404666
store_and_fwd_flag       404666
PULocationID                  0
DOLocationID                  0
payment_type                  0
fare_amount                   0
extra                         0
mta_tax                       0
tip_amount                    0
tolls_amount                  0
improvement_surcharge         0
total_amount                  0
congestion_surcharge     404666
Airport_fee              404666
dtype: int64

Observamos una gran cantidad de nulos representando 404mil viajes donde no se registró cantidad de pasajeros, flag, ni otros cargos. Sin embargo datos importantes como distancia y tarifa están presentes

In [24]:
filas_con_nulos = df1[df1.isnull().any(axis=1)]
filas_con_nulos


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
3319167         2  2024-05-01 00:46:00   2024-05-01 00:54:00              NaN   
3319168         2  2024-05-01 00:23:22   2024-05-01 00:35:48              NaN   
3319169         2  2024-05-01 00:33:56   2024-05-01 00:35:50              NaN   
3319170         1  2024-05-01 00:01:47   2024-05-01 00:10:00              NaN   
3319171         2  2024-05-01 00:45:56   2024-05-01 00:52:02              NaN   
...           ...                  ...                   ...              ...   
3723828         2  2024-05-31 23:45:00   2024-06-01 00:09:00              NaN   
3723829         2  2024-05-31 23:10:52   2024-05-31 23:38:05              NaN   
3723830         2  2024-05-31 23:23:57   2024-05-31 23:38:35              NaN   
3723831         2  2024-05-31 23:44:12   2024-05-31 23:47:44              NaN   
3723832         2  2024-05-31 23:11:10   2024-05-31 23:38:27              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
3319167           1.65         NaN               None           186   
3319168           2.42         NaN               None           170   
3319169           0.40         NaN               None           238   
3319170           1.10         NaN               None           113   
3319171           2.59         NaN               None           112   
...                ...         ...                ...           ...   
3723828           4.07         NaN               None            88   
3723829           4.81         NaN               None           231   
3723830           2.03         NaN               None           142   
3723831           0.94         NaN               None           141   
3723832           9.25         NaN               None           232   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
3319167            79             0         1.74    0.0      0.5        0.00   
3319168            50             0        17.10    0.0      0.5        3.00   
3319169           238             0         7.13    0.0      0.5        0.00   
3319170           170             0        10.64    0.0      0.5        0.00   
3319171           157             0        11.69    0.0      0.5        1.00   
...               ...           ...          ...    ...      ...         ...   
3723828           186             0        29.93    0.0      0.5        0.00   
3723829           143             0        30.71    0.0      0.5        0.00   
3723830           140             0        17.15    0.0      0.5        4.23   
3723831           263             0        -1.00    0.0      0.5        0.00   
3723832            42             0        33.65    0.0      0.5        0.00   

         tolls_amount  improvement_surcharge  total_amount  \
3319167           0.0                    1.0          5.74   
3319168           0.0                    1.0         24.10   
3319169           0.0                    1.0         11.13   
3319170           0.0                    1.0         14.64   
3319171           0.0                    1.0         14.19   
...               ...                    ...           ...   
3723828           0.0                    1.0         33.93   
3723829           0.0                    1.0         34.71   
3723830           0.0                    1.0         25.38   
3723831           0.0                    1.0          3.00   
3723832           0.0                    1.0         37.65   

         congestion_surcharge  Airport_fee  
3319167                   NaN          NaN  
3319168                   NaN          NaN  
3319169                   NaN          NaN  
3319170                   NaN          NaN  
3319171                   NaN          NaN  
...                       ...          ...  
3723828                   NaN          NaN  
3723829                   NaN          NaN  
3723830                   NaN          NaN  
3723831                   N

Observamos 404666 registros en varias columnas nulas, debido a que son información que ingresa el conductor manualmente en el recorrido

In [28]:
duplicated_rows = df1[df1.duplicated()]
duplicated_rows

Empty DataFrame
Columns: [VendorID, lpep_pickup_datetime, lpep_dropoff_datetime, store_and_fwd_flag, RatecodeID, PULocationID, DOLocationID, passenger_count, trip_distance, fare_amount, extra, mta_tax, tip_amount, tolls_amount, ehail_fee, improvement_surcharge, total_amount, payment_type, trip_type, congestion_surcharge]
Index: []

No hay datos duplicados. Se deciden conservar las 19 columnas por su utilidad para el análisis

## 2. Archivo green_tripdate_2024_05.parquet

In [4]:
df2 = pd.read_parquet(r'C:\Users\migue\OneDrive\Escritorio\Datasets Taxi\green_tripdata_2024-05.parquet', engine='pyarrow')
df2

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0             2  2024-05-01 00:07:08   2024-05-01 00:15:03                  N   
1             2  2024-05-01 00:30:48   2024-05-01 00:35:49                  N   
2             2  2024-05-01 00:34:13   2024-05-01 00:38:07                  N   
3             2  2024-05-01 00:58:01   2024-05-01 01:14:41                  N   
4             2  2024-05-01 00:11:45   2024-05-01 00:20:38                  N   
...         ...                  ...                   ...                ...   
60998         2  2024-05-31 22:20:00   2024-05-31 22:34:00               None   
60999         2  2024-05-31 22:08:00   2024-05-31 22:22:00               None   
61000         2  2024-05-31 22:24:00   2024-05-31 22:46:00               None   
61001         2  2024-05-31 23:27:14   2024-05-31 23:52:15               None   
61002         2  2024-05-31 23:30:00   2024-05-31 23:44:00               None   

       RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
0             1.0            65            49              1.0           1.24   
1             1.0             7           179              1.0           0.94   
2             1.0            74            42              1.0           0.84   
3             1.0            75           235              1.0           6.07   
4             1.0           256            49              2.0           2.06   
...           ...           ...           ...              ...            ...   
60998         NaN           166           239              NaN           2.26   
60999         NaN           149            89              NaN           2.75   
61000         NaN            89            97              NaN           4.29   
61001         NaN            80            61              NaN           3.53   
61002         NaN           200           243              NaN           5.05   

       fare_amount  extra  mta_tax  tip_amount  tolls_amount  ehail_fee  \
0             9.30    1.0      0.5        2.00          0.00        NaN   
1             7.20    1.0      0.5        1.94          0.00        NaN   
2             6.50    1.0      0.5        0.00          0.00        NaN   
3            25.40    1.0      0.5        5.00          0.00        NaN   
4            12.10    1.0      0.5        2.92          0.00        NaN   
...            ...    ...      ...         ...           ...        ...   
60998        17.22    0.0      0.0        4.09          0.00        NaN   
60999        17.39    0.0      0.0        0.89          0.00        NaN   
61000        30.30    0.0      0.0        1.54          0.00        NaN   
61001        23.02    0.0      0.0        0.00          0.00        NaN   
61002        23.01    0.0      0.0        5.00          3.18        NaN   

       improvement_surcharge  total_amount  payment_type  trip_type  \
0                        1.0         13.80           1.0        1.0   
1                        1.0         11.64           1.0        1.0   
2                        1.0          9.00           2.0        1.0   
3                        1.0         32.90           1.0        1.0   
4                        1.0         17.52           1.0        1.0   
...                      ...           ...           ...        ...   
60998                    1.0         25.06           NaN        NaN   
60999                    1.0         19.28           NaN        NaN   
61000                    1.0         32.84           NaN        NaN   
61001                    1.0         24.02           NaN        NaN   
61002                    1.0         32.19           NaN        NaN   

       congestion_surcharge  
0                       0.0  
1                       0.0  
2                       0.0  
3                       0.0  
4                       0.0  
...                     ...  
60998                   NaN  
60999                   NaN  
61000                   NaN  
61001  

In [5]:
duplicated_rows = df2[df2.duplicated()]
duplicated_rows

Empty DataFrame
Columns: [VendorID, lpep_pickup_datetime, lpep_dropoff_datetime, store_and_fwd_flag, RatecodeID, PULocationID, DOLocationID, passenger_count, trip_distance, fare_amount, extra, mta_tax, tip_amount, tolls_amount, ehail_fee, improvement_surcharge, total_amount, payment_type, trip_type, congestion_surcharge]
Index: []

No hay duplicados

In [6]:
df2.isna().sum()

VendorID                     0
lpep_pickup_datetime         0
lpep_dropoff_datetime        0
store_and_fwd_flag        1892
RatecodeID                1892
PULocationID                 0
DOLocationID                 0
passenger_count           1892
trip_distance                0
fare_amount                  0
extra                        0
mta_tax                      0
tip_amount                   0
tolls_amount                 0
ehail_fee                61003
improvement_surcharge        0
total_amount                 0
payment_type              1892
trip_type                 1906
congestion_surcharge      1892
dtype: int64

Se elimina la columna ehail_fee por ser nula en toda la data, posiblemente un dato que no se está registrando aún. Vemos que los otros nulos son de columnas de baja relevancia

In [8]:
df2 = df2.drop('ehail_fee', axis=1)

In [9]:

filas_con_nulos = df2[df2.isnull().any(axis=1)]
filas_con_nulos

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
5383          1  2024-05-03 16:49:58   2024-05-03 18:00:37                  N   
5712          1  2024-05-03 18:23:47   2024-05-03 19:21:13                  N   
8593          1  2024-05-05 12:56:11   2024-05-05 14:01:29                  N   
12985         1  2024-05-07 18:02:11   2024-05-07 18:02:11                  Y   
16024         1  2024-05-09 10:06:57   2024-05-09 11:07:41                  N   
...         ...                  ...                   ...                ...   
60998         2  2024-05-31 22:20:00   2024-05-31 22:34:00               None   
60999         2  2024-05-31 22:08:00   2024-05-31 22:22:00               None   
61000         2  2024-05-31 22:24:00   2024-05-31 22:46:00               None   
61001         2  2024-05-31 23:27:14   2024-05-31 23:52:15               None   
61002         2  2024-05-31 23:30:00   2024-05-31 23:44:00               None   

       RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
5383         99.0            37           241              1.0          11.80   
5712         99.0           254           177              1.0          14.70   
8593         99.0           179            92              1.0           5.80   
12985        99.0            75           264              0.0           0.00   
16024        99.0           167           237              1.0           5.00   
...           ...           ...           ...              ...            ...   
60998         NaN           166           239              NaN           2.26   
60999         NaN           149            89              NaN           2.75   
61000         NaN            89            97              NaN           4.29   
61001         NaN            80            61              NaN           3.53   
61002         NaN           200           243              NaN           5.05   

       fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
5383         46.50    0.0      1.5        0.00          0.00   
5712         53.50    0.0      1.5        0.00          0.00   
8593         28.50    0.0      1.5        0.00          0.00   
12985        81.94    0.0      0.0        0.00          0.00   
16024        30.50    0.0      1.5        0.00          0.00   
...            ...    ...      ...         ...           ...   
60998        17.22    0.0      0.0        4.09          0.00   
60999        17.39    0.0      0.0        0.89          0.00   
61000        30.30    0.0      0.0        1.54          0.00   
61001        23.02    0.0      0.0        0.00          0.00   
61002        23.01    0.0      0.0        5.00          3.18   

       improvement_surcharge  total_amount  payment_type  trip_type  \
5383                     1.0         48.00           1.0        NaN   
5712                     1.0         55.00           1.0        NaN   
8593                     1.0         30.00           1.0        NaN   
12985                    0.0         81.94           3.0        NaN   
16024                    1.0         32.00           1.0        NaN   
...                      ...           ...           ...        ...   
60998                    1.0         25.06           NaN        NaN   
60999                    1.0         19.28           NaN        NaN   
61000                    1.0         32.84           NaN        NaN   
61001                    1.0         24.02           NaN        NaN   
61002                    1.0         32.19           NaN        NaN   

       congestion_surcharge  
5383                    0.0  
5712                    0.0  
8593                    0.0  
12985                   0.0  
16024                   0.0  
...                     ...  
60998                   NaN  
60999                   NaN  
61000                   NaN  
61001                   NaN  
61002                   NaN  

[1906 rows x 19 columns]

Existe una gran cantidad relativa de nulos en ciertas columnas, se decide dejar como dato faltante y no afectar el dataset

## Archivo For-hire vehicle datatrip 2024 05

In [30]:
df3 = pd.read_parquet(r'C:\Users\migue\OneDrive\Escritorio\Datasets Taxi\fhv_tripdata_2024-05.parquet', engine='pyarrow')
df3

dispatching_base_num     pickup_datetime    dropOff_datetime  \
0                     B00008 2024-05-01 00:53:00 2024-05-01 02:08:00   
1                     B00009 2024-05-01 00:13:00 2024-05-01 00:33:00   
2                     B00009 2024-05-01 00:34:00 2024-05-01 00:56:00   
3                     B00111 2024-05-01 00:01:00 2024-05-01 01:20:00   
4                     B00111 2024-05-01 00:34:01 2024-05-01 01:29:00   
...                      ...                 ...                 ...   
1352497               B03492 2024-05-31 23:10:48 2024-05-31 23:11:07   
1352498               B03492 2024-05-31 23:30:04 2024-05-31 23:37:30   
1352499               B03512 2024-05-31 23:42:07 2024-05-31 23:58:01   
1352500               B03529 2024-05-31 23:02:35 2024-05-31 23:15:30   
1352501               B03572 2024-05-31 23:24:33 2024-05-31 23:53:25   

         PUlocationID  DOlocationID  SR_Flag Affiliated_base_number  
0                 NaN           NaN      NaN                 B00008  
1                 NaN           NaN      NaN                 B00009  
2                 NaN           NaN      NaN                 B00009  
3                 NaN           NaN      NaN                 B00111  
4                 NaN           NaN      NaN                 B00111  
...               ...           ...      ...                    ...  
1352497           NaN          11.0      NaN                 B03492  
1352498           NaN          67.0      NaN                 B03492  
1352499           NaN          36.0      NaN                 B03512  
1352500           NaN          37.0      NaN                 B03404  
1352501          56.0         130.0      NaN        B03572           

[1352502 rows x 7 columns]

In [31]:
df3.isna().sum()

dispatching_base_num            0
pickup_datetime                 0
dropOff_datetime                0
PUlocationID              1075862
DOlocationID               195147
SR_Flag                   1352502
Affiliated_base_number          0
dtype: int64

vemos gran cantidad de nulos en las ID de locación de subida y bajada de pasajeros

## For-hire Vehicle High Volumen (fhvhv_tripdata_2024_05)

Archivo muy pesado para esta plataforma, se conectará y analizará directamente en el datawarehouse